ambry-store/src/main/java/com.github.ambry.store/StoreMessageReadSet.java

```diff
@@ -171,7 +198,21 @@ public long writeTo(int index, WritableByteChannel channel, long relativeOffset,
     long startOffset = options.getOffset() + relativeOffset;
     long sizeToRead = Math.min(maxSize, options.getMessageInfo().getSize() - relativeOffset);
     logger.trace("Blob Message Read Set position {} count {}", startOffset, sizeToRead);
-    long written = options.getChannel().transferTo(startOffset, sizeToRead, channel);
+    long written = 0;
+    long bufStartOffset = relativeOffset - options.getPrefetchedDataRelativeOffset();
```

Comment:
Would prefer if we can completely isolate the `transferTo()` part from the prefetch part. Basically, not comfortable with `relativeOffset - options.getPrefetchedDataRelativeOffset()` being computed if the latter is an invalid value (-1).

wherever the commonality ends, do you think it is better to rewrite this as:
```java
if (options.getPrefetchedDataRelativeOffset() == -1) {
  ...
  // transferTo()
  ...
} else {
  ...
  // read prefetched buffer
  ...
}
```

Reply:
That's what I do initially, but as code goes to PROD, I think it's safe to fall back to transferTo if prefetch fails for some unknown reasons. After we figure out performance issue and don't see warning message in prod, we can isolate them.

Reply:
if the prefetch fails, wouldn't you know it in advance (when the prefetch was attempted?). So could you not maintain that `prefetchedDataRelativeOffset` is -1?